ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/FSharpIdentifierTooltipProvider.fs

```diff
@@ -43,61 +107,59 @@ type FSharpIdentifierTooltipProvider(lifetime, solution, presenter, xmlDocServic
         match document.GetPsiSourceFile(solution) with
-        | null -> String.Empty
-        | sourceFile when not (sourceFile.IsValid()) -> String.Empty
+        | null -> emptyPresentation
+        | sourceFile when not (sourceFile.IsValid()) -> emptyPresentation
         | sourceFile ->
 
         let documentRange = DocumentRange(document, highlighter.Range)
         match x.GetPsiFile(sourceFile, documentRange).As<IFSharpFile>() with
-        | null -> String.Empty
+        | null -> emptyPresentation
         | fsFile ->
 
         match fsFile.FindTokenAt(documentRange.StartOffset).As<FSharpIdentifierToken>() with
-        | null -> String.Empty
+        | null -> emptyPresentation
         | token ->
 
         match fsFile.GetParseAndCheckResults(true, opName) with
-        | None -> String.Empty
+        | None -> emptyPresentation
         | Some results ->
 
-        let result = List()
         let (FSharpToolTipText layouts) = FSharpIdentifierTooltipProvider.GetFSharpToolTipText(opName, results.CheckResults, token)
-        
-        layouts |> List.iter (function
+
+        layouts
+        |> List.collect (function
             | FSharpStructuredToolTipElement.None
-            | FSharpStructuredToolTipElement.CompositionError _ -> ()
+            | FSharpStructuredToolTipElement.CompositionError _ -> []
 
             | FSharpStructuredToolTipElement.Group(overloads) ->
-                overloads |> List.iter (fun overload ->
+                overloads
+                |> List.map (fun overload ->
                     [ if not (isEmptyL overload.MainDescription) then
-                          yield showL overload.MainDescription
+                          yield taggedTextToRichText overload.MainDescription
 
                       if not overload.TypeMapping.IsEmpty then
-                          yield overload.TypeMapping |> List.map showL |> String.concat "\n"
+                          yield overload.TypeMapping |> List.map taggedTextToRichText |> richTextJoin "\n"
 
                       match xmlDocService.GetXmlDoc(overload.XmlDoc) with
                       | null -> ()
                       | xmlDocText when xmlDocText.Text.IsNullOrWhitespace() -> ()
-                      | xmlDocText -> yield xmlDocText.Text
+                      | xmlDocText -> yield xmlDocText.RichText
 
                       match overload.Remarks with
                       | Some remarks when not (isEmptyL remarks) ->
-                          yield showL remarks
+                          yield taggedTextToRichText remarks
                       | _ -> () ]
-                    |> String.concat "\n\n"
-                    |> result.Add))
-
-        result.Join(RiderTooltipSeparator)
-
-    override x.GetRichTooltip(highlighter) =
-        RichTextBlock(x.GetTooltip(highlighter))
+                    |> richTextJoin "\n\n"))
+        // todo: add support for _RIDER_HORIZONTAL_LINE_TOOLTIP_SEPARATOR_
```

Comment:
@kirillgla Is it supposed to work with the rich text tooltips?

Reply:
Yes, it is. I tried it out locally, and I can say that replacing `"\n--------------------\n"` with `"_RIDER_HORIZONTAL_LINE_TOOLTIP_SEPARATOR_"` works just like one would expect it to
![image](https://user-images.githubusercontent.com/31822295/93818620-481c5680-fc63-11ea-9b62-3d44a3f93f7f.png)

Reply:
Thanks! 🙂